html/user/openssl_sign_action.php

```diff
@@ -0,0 +1,134 @@
+<?php
+// This file is part of BOINC.
+// http://boinc.berkeley.edu
+// Copyright (C) 2008 University of California
+//
+// BOINC is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// BOINC is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+
+require_once("../inc/boinc_db.inc");
+require_once("../inc/user.inc");
+require_once("../inc/util.inc");
+
+// Needs to be defined, unknown use.
+check_get_args(array("tnow", "ttok"));
+
+// Check the user is online
+$user = get_logged_in_user();
+check_tokens($user->authenticator);
+
+// For some reason, I'm only able to use numbers, no letters.
+// Need to provide security, but full functionality is likewise desirable.
+$user_data = sanitize_tags(post_str("user_data", true));
+
+// Following line a potential bug.
+// $user_data = BoincDb::escape_string($user_data);
+
+//Has the user entered data with length > 0 & <= 256?
+if (($user_data) && (strlen($user_data) <= 256)) {
+	// Retrieve the logged in user's UserID.
+	$userid = $user->id;
+
+	// Create the message which will be signed.
+	$message_data = "$userid $user_data";
+
+	// Create a digest of the message_data.
+	// Without the digest, the openssl_sign function fails to sign lower case characters.
```

Comment:
Woah, wait. Really? How can (hybrid) signing function have problems with some letters? This needs to be investigated and the double hashing step removed.

Reply:
Without creating a message digest from the user data to feed into openssl_sign, any user data which included lower-case data would change the text file output to unrecognizable garbage like:

```
2d2d 2d2d 2d42 4547 494e 2055 5345 5249
4420 5645 5249 4649 4341 5449 4f4e 2d2d
2d2d 2d0a 5573 6572 4944 3a20 310a 5573
6572 4461 7461 3a20 2231 3233 3132 330a
4d65 7373 6167 653a 2031 2022 3132 3331
3233 0a53 6967 6e61 7475 7265 3a20 71ce
1029 7aa8 b0d6 221f 5a83 29b1 4962 3092
80b9 71a1 fce4 17ce 35e1 00d6 7b6a 3ac7
888b bbb2 361e 3ee7 982f 578d eb5d cda8
393e 0bf5 eee7 5b1b 1def 1f10 b545 678c
51ca 8d3e b117 df08 12df 2449 a6fc 646f
abbe 3b86 f6db 9220 77e6 ea4f 8947 628c
3f31 d640 5b9c c987 0771 423f 9d27 c296
a7ca 1783 c6f2 0b99 6b18 6964 aad5 1657
4929 5daa 8a1a ef27 5534 1c7b 0dda ad1f
9860 943b 4f84 0d62 5fe4 44b8 7289 6098
26b1 545c baa9 78c2 d187 0c4b c814 a53b
bc96 b5de 0866 0f9f 5e7a 217a b142 675f
b8bc bd26 10d0 f429 5498 9976 b002 7d06
6689 e30c 651d f68f b7f5 8498 c638 b85b
7f4e b7b1 8d60 49ea 38a8 ddcf 2f80 5515
191c 7d3a 0ac8 7459 029f e57c 8638 0a56
6572 6966 6963 6174 696f 6e3a 2047 4f4f
440a 2d2d 2d2d 2d45 4e44 2055 5345 5249
4420 5645 5249 4649 4341 5449 4f4e 2d2d
2d2d 2d
```

Using a message digest prevents this issue completely, I'm unsure why the above issue occurs but using a message digest seems to be an appropriate solution.

Reply:
Using message digest is not a solution, it is a workaround. And that output looks really, really bad. This is a serious issue. You got me hooked, I will test myself.

Reply:
Looks like I was running an old version of OpenSSL: https://github.com/marius311/boinc-server-docker/issues/26#issuecomment-330553493

Could explain the glitch..

Reply:
I've managed to upgrade the openssl library to  1.0.1t, but can't upgrade any further on debian jessie despite the openssl website stating that it's unsupported now..

With this version the removal of the message digest workaround is not required.